File: minigit-3/py/commands/show.py
"""show command - Show object content."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Show object content."""
    from utils import find_repo_root, read_object
    from refs import resolve_ref, get_head_sha
    from objects import read_commit, read_tree, read_blob, read_tag
    from diff_algo import diff_files

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    target = 'HEAD'
    if args:
        target = args[0]

    # Handle path specifier (e.g., HEAD:file.txt)
    if ':' in target:
        ref_part, path = target.split(':', 1)
        ref_sha = resolve_ref(repo_root, ref_part)
        if ref_sha is None:
            print(f"error: unknown revision '{ref_part}'", file=sys.stderr)
            return 1

        # Get the file content
        commit = read_commit(repo_root, ref_sha)
        entry = get_tree_entry(repo_root, commit['tree'], path)
        if entry is None:
            print(f"error: path '{path}' not found in '{ref_part}'", file=sys.stderr)
            return 1

        content = read_blob(repo_root, entry.sha)
        try:
            print(content.decode(), end='')
        except:
            sys.stdout.buffer.write(content)
        return 0

    # Resolve the reference
    sha = resolve_ref(repo_root, target)
    if sha is None:
        print(f"error: unknown revision '{target}'", file=sys.stderr)
        return 1

    # Determine object type
    obj_type, data = read_object(repo_root, sha)

    if obj_type == 'commit':
        show_commit(repo_root, sha)
    elif obj_type == 'tree':
        show_tree(repo_root, sha)
    elif obj_type == 'blob':
        try:
            print(data.decode(), end='')
        except:
            sys.stdout.buffer.write(data)
    elif obj_type == 'tag':
        show_tag(repo_root, sha)
    else:
        print(f"Unknown object type: {obj_type}", file=sys.stderr)
        return 1

    return 0


def get_tree_entry(repo_root, tree_sha: str, path: str):
    """Get entry from tree by path."""
    from objects import read_tree

    parts = path.split('/')
    current_sha = tree_sha

    for i, part in enumerate(parts):
        entries = read_tree(repo_root, current_sha)
        found = None
        for entry in entries:
            if entry.name == part:
                found = entry
                break

        if found is None:
            return None

        if i < len(parts) - 1:
            if found.mode != '40000':
                return None
            current_sha = found.sha
        else:
            return found

    return None


def show_commit(repo_root, sha: str):
    """Display commit information."""
    from objects import read_commit, read_tree, read_blob
    from datetime import datetime
    from diff_algo import diff_files
    from utils import is_binary_file

    commit = read_commit(repo_root, sha)

    print(f"commit {sha}")

    # Parse author
    author = commit.get('author', 'Unknown')
    parts = author.rsplit(' ', 2)
    if len(parts) >= 3:
        author_info = parts[0]
        timestamp = parts[1]
        tz = parts[2]

        try:
            ts = int(timestamp)
            dt = datetime.utcfromtimestamp(ts)
            date_str = dt.strftime('%a %b %d %H:%M:%S %Y') + f" {tz}"
        except:
            date_str = f"{timestamp} {tz}"

        print(f"Author: {author_info}")
        print(f"Date:   {date_str}")
    else:
        print(f"Author: {author}")

    print()
    for line in commit['message'].split('\n'):
        print(f"    {line}")
    print()

    # Show diff from parent
    if commit['parents']:
        parent_sha = commit['parents'][0]
        parent_commit = read_commit(repo_root, parent_sha)
        show_tree_diff(repo_root, parent_commit['tree'], commit['tree'])
    else:
        # First commit - show all files as added
        show_tree_diff(repo_root, None, commit['tree'])


def show_tree_diff(repo_root, old_tree_sha, new_tree_sha):
    """Show diff between two trees."""
    from objects import read_tree, read_blob
    from diff_algo import diff_files
    from utils import is_binary_file

    def get_tree_files(tree_sha, prefix=''):
        if tree_sha is None:
            return {}
        files = {}
        try:
            tree_entries = read_tree(repo_root, tree_sha)
            for entry in tree_entries:
                path = f"{prefix}{entry.name}" if prefix else entry.name
                if entry.mode == '40000':
                    files.update(get_tree_files(entry.sha, path + '/'))
                else:
                    files[path] = entry.sha
        except:
            pass
        return files

    old_files = get_tree_files(old_tree_sha)
    new_files = get_tree_files(new_tree_sha)

    all_paths = set(old_files.keys()) | set(new_files.keys())

    for path in sorted(all_paths):
        old_sha = old_files.get(path)
        new_sha = new_files.get(path)

        if old_sha == new_sha:
            continue

        old_content = None
        new_content = None

        if old_sha:
            try:
                data = read_blob(repo_root, old_sha)
                if is_binary_file(data):
                    print(f"Binary file {path} differs")
                    continue
                old_content = data.decode()
            except:
                pass

        if new_sha:
            try:
                data = read_blob(repo_root, new_sha)
                if is_binary_file(data):
                    print(f"Binary file {path} differs")
                    continue
                new_content = data.decode()
            except:
                pass

        diff_output = diff_files(old_content, new_content, path)
        if diff_output:
            print(diff_output)


def show_tree(repo_root, sha: str):
    """Display tree contents."""
    from objects import read_tree

    entries = read_tree(repo_root, sha)
    for entry in entries:
        type_name = 'tree' if entry.mode == '40000' else 'blob'
        print(f"{entry.mode} {type_name} {entry.sha}\t{entry.name}")


def show_tag(repo_root, sha: str):
    """Display tag information."""
    from objects import read_tag

    tag = read_tag(repo_root, sha)

    print(f"tag {tag['tag']}")
    print(f"Tagger: {tag['tagger']}")
    print()
    print(tag['message'])
    print()
    print(f"commit {tag['object']}")

File: minigit-3/ts/src/commands/show.ts
// show command - Show object content

import { findRepoRoot, shortSha, formatDate, parseTimestamp } from '../utils';
import { resolveRef, getHeadCommit } from '../refs';
import { readObject, parseCommit, parseTree, parseTag as parseTagObject } from '../objects';
import { diffFiles, formatDiff } from '../diff-algo';

export function show(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  const objectRef = args[0] || 'HEAD';

  // Check for path syntax: <commit>:<path>
  if (objectRef.includes(':')) {
    showObjectAtPath(objectRef, repoRoot);
    return;
  }

  const sha = resolveRef(objectRef, repoRoot);
  if (!sha) {
    console.error(`fatal: bad object ${objectRef}`);
    process.exit(1);
  }

  const obj = readObject(sha, repoRoot);

  switch (obj.type) {
    case 'commit':
      showCommit(sha, obj.content, repoRoot);
      break;
    case 'tree':
      showTree(sha, obj.content, repoRoot);
      break;
    case 'blob':
      console.log(obj.content.toString());
      break;
    case 'tag':
      showTag(sha, obj.content, repoRoot);
      break;
    default:
      console.log(obj.content.toString());
  }
}

function showCommit(sha: string, content: Buffer, repoRoot: string): void {
  const commit = parseCommit(content);

  console.log(`commit ${sha}`);

  if (commit.parents.length > 1) {
    console.log(`Merge: ${commit.parents.map(p => shortSha(p)).join(' ')}`);
  }

  // Parse author info
  const authorMatch = commit.author.match(/^(.+?) <(.+?)> (\d+) ([+-]\d{4})$/);
  if (authorMatch) {
    const [, name, email, timestamp, tz] = authorMatch;
    const dateStr = formatDate(parseInt(timestamp, 10), tz);
    console.log(`Author: ${name} <${email}>`);
    console.log(`Date:   ${dateStr}`);
  } else {
    console.log(`Author: ${commit.author}`);
  }

  console.log('');
  for (const line of commit.message.split('\n')) {
    console.log(`    ${line}`);
  }
  console.log('');

  // Show diff from parent
  if (commit.parents.length > 0) {
    const parentSha = commit.parents[0];
    showCommitDiff(parentSha, sha, repoRoot);
  } else {
    // Initial commit - show all files as added
    showInitialCommitDiff(commit.tree, repoRoot);
  }
}

function showCommitDiff(parentSha: string, commitSha: string, repoRoot: string): void {
  const parentCommit = parseCommit(readObject(parentSha, repoRoot).content);
  const commit = parseCommit(readObject(commitSha, repoRoot).content);

  const parentFiles = new Map<string, { sha: string; mode: string }>();
  const commitFiles = new Map<string, { sha: string; mode: string }>();

  collectTreeFilesWithMode(parentCommit.tree, '', repoRoot, parentFiles);
  collectTreeFilesWithMode(commit.tree, '', repoRoot, commitFiles);

  const allPaths = new Set([...parentFiles.keys(), ...commitFiles.keys()]);

  for (const name of [...allPaths].sort()) {
    const parentEntry = parentFiles.get(name);
    const commitEntry = commitFiles.get(name);

    if (!parentEntry && commitEntry) {
      // New file
      const obj = readObject(commitEntry.sha, repoRoot);
      const diff = diffFiles('', obj.content.toString(), {
        oldPath: name,
        newPath: name,
        newMode: commitEntry.mode
      });
      diff.isNew = true;
      console.log(formatDiff(diff));
    } else if (parentEntry && !commitEntry) {
      // Deleted file
      const obj = readObject(parentEntry.sha, repoRoot);
      const diff = diffFiles(obj.content.toString(), '', {
        oldPath: name,
        newPath: name,
        oldMode: parentEntry.mode
      });
      diff.isDeleted = true;
      console.log(formatDiff(diff));
    } else if (parentEntry && commitEntry && parentEntry.sha !== commitEntry.sha) {
      // Modified file
      const parentObj = readObject(parentEntry.sha, repoRoot);
      const commitObj = readObject(commitEntry.sha, repoRoot);
      const diff = diffFiles(parentObj.content.toString(), commitObj.content.toString(), {
        oldPath: name,
        newPath: name,
        oldMode: parentEntry.mode,
        newMode: commitEntry.mode
      });
      console.log(formatDiff(diff));
    }
  }
}

function showInitialCommitDiff(treeSha: string, repoRoot: string): void {
  const files = new Map<string, { sha: string; mode: string }>();
  collectTreeFilesWithMode(treeSha, '', repoRoot, files);

  for (const [name, entry] of [...files].sort((a, b) => a[0].localeCompare(b[0]))) {
    const obj = readObject(entry.sha, repoRoot);
    const diff = diffFiles('', obj.content.toString(), {
      oldPath: name,
      newPath: name,
      newMode: entry.mode
    });
    diff.isNew = true;
    console.log(formatDiff(diff));
  }
}

function showTree(sha: string, content: Buffer, repoRoot: string): void {
  const entries = parseTree(content);

  for (const entry of entries) {
    const type = entry.mode.startsWith('40') ? 'tree' : 'blob';
    console.log(`${entry.mode} ${type} ${entry.sha}\t${entry.name}`);
  }
}

function showTag(sha: string, content: Buffer, repoRoot: string): void {
  const tag = parseTagObject(content);

  console.log(`tag ${tag.tag}`);
  console.log(`Tagger: ${tag.tagger}`);
  console.log('');
  console.log(tag.message);
  console.log('');

  // Show referenced object
  const obj = readObject(tag.object, repoRoot);
  if (obj.type === 'commit') {
    showCommit(tag.object, obj.content, repoRoot);
  }
}

function showObjectAtPath(ref: string, repoRoot: string): void {
  const colonIndex = ref.indexOf(':');
  const commitRef = ref.slice(0, colonIndex);
  const pathPart = ref.slice(colonIndex + 1);

  const sha = resolveRef(commitRef, repoRoot);
  if (!sha) {
    console.error(`fatal: bad revision '${commitRef}'`);
    process.exit(1);
  }

  const obj = readObject(sha, repoRoot);
  if (obj.type !== 'commit') {
    console.error(`fatal: '${commitRef}' is not a commit`);
    process.exit(1);
  }

  const commit = parseCommit(obj.content);
  const files = new Map<string, { sha: string; mode: string }>();
  collectTreeFilesWithMode(commit.tree, '', repoRoot, files);

  const entry = files.get(pathPart);
  if (!entry) {
    console.error(`fatal: path '${pathPart}' does not exist in '${commitRef}'`);
    process.exit(1);
  }

  const blobObj = readObject(entry.sha, repoRoot);
  console.log(blobObj.content.toString());
}

function collectTreeFilesWithMode(treeSha: string, prefix: string, repoRoot: string, files: Map<string, { sha: string; mode: string }>): void {
  const treeObj = readObject(treeSha, repoRoot);
  const entries = parseTree(treeObj.content);

  for (const entry of entries) {
    const name = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.type === 'tree') {
      collectTreeFilesWithMode(entry.sha, name, repoRoot, files);
    } else {
      files.set(name, { sha: entry.sha, mode: entry.mode });
    }
  }
}
